src/Compilers/CSharp/Portable/BoundTree/TupleBinaryOperatorInfo.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// A tree of binary operators for tuple comparisons.
+    ///
+    /// For `(a, (b, c)) == (d, (e, f))` we'll hold a Multiple with two elements.
+    /// The first element is a Single (describing the binary operator and conversions are involved in `a == d`).
+    /// The second element is a Multiple containing two Singles (one for the `b == e` comparison and the other for `c == f`).
+    /// </summary>
+    internal abstract class TupleBinaryOperatorInfo
+    {
+        internal abstract bool IsSingle();
+        internal readonly TypeSymbol LeftConvertedType;
+        internal readonly TypeSymbol RightConvertedType;
+#if DEBUG
+        internal abstract TreeDumperNode DumpCore();
+        internal string Dump() => TreeDumper.DumpCompact(DumpCore());
+#endif
+
+        private TupleBinaryOperatorInfo(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType)
+        {
+            LeftConvertedType = leftConvertedType;
+            RightConvertedType = rightConvertedType;
+        }
+
+        /// <summary>
+        /// Holds the information for an element-wise comparison (like `a == b`)
+        /// </summary>
+        internal class Single : TupleBinaryOperatorInfo
+        {
+            internal readonly BinaryOperatorKind Kind;
+            internal readonly MethodSymbol MethodSymbolOpt; // User-defined comparison operator, if applicable
+
+            // To convert the result of comparison to bool
+            internal readonly Conversion BoolConversion;
+            internal readonly UnaryOperatorSignature BoolOperator; // Information for op_true or op_false
```

Comment:
>UnaryOperatorSignature [](start = 30, length = 22)

Why does this have to be UnaryOperatorSignature, what does it gives us that we cannot get from the underlying MethodSymbol?

Reply:
We end up using all four parts (`Kind`, `OperandType`, `ReturnType` and `Method`) from the signature.
That follows the same logic as binding a regular operator, which uses `OperandType` and then returns `new BoundUnaryOperator(node, signature.Kind, resultOperand, ConstantValue.NotAvailable, signature.Method, resultKind, originalUserDefinedOperators, signature.ReturnType)` (using the other 3 parts).

`ReturnType` and `OperandType` could be probably be recovered out of `Method` (I don't know if there are complexities involved).
If we want to slim down the signature object (only keep `Kind` and `Method`), we can file an issue on that. 

---
In reply to: [170489381](https://github.com/dotnet/roslyn/pull/23356#discussion_r170489381) [](ancestors = 170489381)